### PR TITLE
Refactor upgrade strategy to accept null values and use closure stacking

### PR DESCRIPTION
**Pull request by pchojnacki**
Description: see title

/cc: @sqreek @harnash

### CODE REVIEW

File: src/main/groovy/com/wikia/gradle/marathon/common/Marathon.groovy [hunk 1]
```diff
@@ -21,14 +25,15 @@ class Marathon implements Validating {
...
+        this.rawUpgradeStrategy = dslToParamClosure(upgradeStrategy)
     }
 
-    def resolveUpgradeStrategy() {
-        UpgradeStrategy realUpgradeStrategy = new UpgradeStrategy()
-        this.upgradeStrategy.delegate = realUpgradeStrategy
-        this.upgradeStrategy.run()
-        return realUpgradeStrategy
+    UpgradeStrategy resolveUpgradeStrategy() {
+        resolveNullConfig(this.rawUpgradeStrategy, UpgradeStrategy)
```
Comment (sqreek) [inline src/main/groovy/com/wikia/gradle/marathon/common/Marathon.groovy]: here is no return. Is this correct?

File: src/main/groovy/com/wikia/gradle/marathon/common/ConfigResolver.groovy [hunk 2]
```diff
@@ -11,11 +12,19 @@ class ConfigResolver {
         }
     }
 
-    static <T> T resolveConfig(Closure<T> closure, Class<T> klass) {
+    static <T> T resolveConfig(Closure<T> closure, Class<T> klass = null) {
         def instance = klass.newInstance()
```
Comment (sqreek) [inline src/main/groovy/com/wikia/gradle/marathon/common/ConfigResolver.groovy]: If klass now, have default null, is this line create NullPointerException?

**Comment by sqreek** [comment]
and no description here :)

File: src/main/groovy/com/wikia/gradle/marathon/common/Marathon.groovy (hunk 1 above)
Comment (pchojnacki) [inline src/main/groovy/com/wikia/gradle/marathon/common/Marathon.groovy]: groovy has implicit returns

**Comment by sqreek** [comment]
LGTM

**Pull request merged**